cocotb/triggers.py

```diff
@@ -890,7 +891,9 @@ def __repr__(self):
 
 async def with_timeout(trigger, timeout_time, timeout_unit="step"):
     """
-    Waits on triggers, throws an exception if it waits longer than the given time.
+    Waits on triggers and coroutines, throws an exception if it waits longer than the given time.
+
+    When used with a coroutine, the coroutine will be forked and awaited, and the return value forwarded when the coroutine finishes.  If the timeout expires, the coroutine will be killed.
```

Comment:
I think maybe we can blow this out a bit and expand on the actual behaviors of each type of argument: `fork`ed coroutines, `@cocotb.coroutine`-decorated coroutines, waitables and triggers, and `async def` coroutines.

* `fork`ed coroutines have already been started, will wait until the coroutine finishes, and will not be killed after `with_timeout` exits
* `@cocotb.coroutine`-decorated coroutines are started when put in `with_timeout`, but are not killed after `with_timeout` exits
* waitables and triggers will finish or be cancelled
* `async def` coroutines will be started and then killed after exit

Reply:
```suggestion
    Waits on triggers and coroutines, throws an exception if it waits longer than the given time.

    When a :term:`python:coroutine` is passed,
    the callee coroutine is started,
    the caller blocks until the callee completes,
    and the callee's result is returned to the caller.
    If timeout occurs, the callee is killed
    and :exc:`SimTimeoutError` is raised.

    When an unstarted :class:`~cocotb.coroutine`\ is passed,
    the callee coroutine is started,
    the caller blocks until the callee completes,
    and the callee's result is returned to the caller.
    If timeout occurs, the callee `continues to run`
    and :exc:`SimTimeoutError` is raised.

    When a :func:`~cocotb.fork`\ ed coroutine is passed,
    the caller blocks until the callee completes
    and the callee's result is returned to the caller.
    If timeout occurs, the callee `continues to run`
    and :exc:`SimTimeoutError` is raised.

    If a :class:`~cocotb.triggers.Trigger` or :class:`~cocotb.triggers.Waitable` is passed,
    the caller blocks until the trigger fires,
    and the trigger is returned to the caller.
    If timeout occurs, the trigger is cancelled
    and :exc:`SimTimeoutError` is raised.
```